### DISCUSSION TITLE
Database middleware

### DISCUSSION THREAD

**Discussion by fimbault** [routerify/routerify#73] [Q&A] [answered]
Hi, 
Version 2 is neat, great work!
What is the suggested way of connecting to a DB? A shared connection pool variable? 
Thxs

**Comment by gsserge** [comment, marked as answer]
With Routerify you can use any db library that supports `tokio`. Initialize the pool and then pass it as [shared data](https://docs.rs/routerify/2.0.1/routerify/#data-and-state-sharing) to a router instance.
Here's a simple yet fully functional example using [sqlx](https://github.com/launchbadge/sqlx) and Postgres.

#### Cargo.toml
```toml
[dependencies]
tokio = { version = "1", features = ["full"] }
hyper = { version = "0.14", features = ["full"] }
routerify = "2"
sqlx = { version = "0.5", features = [ "runtime-tokio-rustls", "postgres" ] }

```
#### main.rs
```rust
use hyper::{Body, Request, Response, Server};
use routerify::prelude::*;
use routerify::{Router, RouterService};
use sqlx::postgres;
use std::{convert::Infallible, net::SocketAddr};

struct Service {
    db_pool: postgres::PgPool,
}

async fn home(req: Request<Body>) -> Result<Response<Body>, Infallible> {
    // Access the Serice instance.
    let service = req.data::<Service>().unwrap();
    let n_conn = service.db_pool.size();
    Ok(Response::new(format!("{} db connections", n_conn).into()))
}

#[tokio::main]
async fn main() {
    // Create new db pool.
    let pool = postgres::PgPoolOptions::new()
        // Make sure to use correct connection string.
        .connect("postgres://postgres:postgres@localhost/testdb")
        .await
        .expect("unable to create postgresql pool");

    // Create a Service with the db pool.
    let service = Service { db_pool: pool };

    // Pass the Service as shared data to the router; it can be
    // accessed by every handler (inclding error) and middleware.
    let router = Router::builder()
        .data(service)
        .get("/", home)
        .build()
        .unwrap();

    let service = RouterService::new(router).unwrap();

    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    let server = Server::bind(&addr).serve(service);
    println!("App is running on: {}", addr);
    if let Err(err) = server.await {
        eprintln!("Server error: {}", err);
    }
}

```
You don't have to use the `Service` struct btw, passing the pool directly is also fine. It's just my personal preference when building web services using Routerify: define the top-level Service struct to hold config info, logger, db pool and anything else that you may need in the handlers.

**Reply by fimbault** [reply]
Great, thanks a lot, it confirms what I imagined. Might be useful to include that as an example, since sqlx should now work with sqlite also it would do a self contained case, I'll try that. 
Also another use case where static files are served would be useful.

**Reply by gsserge** [reply]
You're welcome!

Your question actually made me think - it might be a good idea to create something like "Routerify by Example" cookbook. Things like db pool, logging, metrics, auth are quite common, and I've got some experience implementing all of them now.

The `examples` folder could be an option, although it might be a bit too heavy on tests and builds if we start expanding dev dependencies with sqlite, slog, prometheus, etc. A separate repo might be a better option.

**Reply by fimbault** [reply]
would be awesome

**Reply by rousan** [reply]
@gsserge , Let's create a separate repo consisting of various examples based on specific use cases.
